File: backend/features/assets/assets-core/src/main/java/de/bennyboer/kicherkrabbe/assets/AssetReference.java
package de.bennyboer.kicherkrabbe.assets;

import lombok.AllArgsConstructor;
import lombok.Value;

import static de.bennyboer.kicherkrabbe.commons.Preconditions.notNull;
import static lombok.AccessLevel.PRIVATE;

@Value
@AllArgsConstructor(access = PRIVATE)
public class AssetReference {

    AssetId assetId;

    AssetReferenceResourceType resourceType;

    AssetResourceId resourceId;

    public static AssetReference of(AssetId assetId, AssetReferenceResourceType resourceType, AssetResourceId resourceId) {
        notNull(assetId, "Asset ID must be given");
        notNull(resourceType, "Resource type must be given");
        notNull(resourceId, "Resource ID must be given");

        return new AssetReference(assetId, resourceType, resourceId);
    }

}

File: frontend/management/src/app/modules/admin/modules/assets/model/asset-reference.ts
import { validateProps } from '@kicherkrabbe/shared';

export class AssetReference {
  readonly resourceType: string;
  readonly resourceId: string;
  readonly resourceName: string;

  private constructor(props: { resourceType: string; resourceId: string; resourceName: string }) {
    validateProps(props);

    this.resourceType = props.resourceType;
    this.resourceId = props.resourceId;
    this.resourceName = props.resourceName;
  }

  static of(props: { resourceType: string; resourceId: string; resourceName: string }): AssetReference {
    return new AssetReference({
      resourceType: props.resourceType,
      resourceId: props.resourceId,
      resourceName: props.resourceName,
    });
  }
}
